java-frontend/src/main/java/org/sonar/java/cfg/CFGLoop.java

```diff
@@ -0,0 +1,178 @@
+    if (answer || isBreak(block)) {
+      blocks.add(block);
+    }
+    return answer;
+  }
+
+  private boolean returnsToStart(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
+    Set<Block> localSuccessors = localSuccessors(block, container);
+    if (localSuccessors == null) {
+      return true;
+    }
+    boolean answer = false;
+    for (CFG.Block successor : localSuccessors) {
+      if (startingBlock == successor) {
+        answer = true;
+      } else {
+        answer |= collectBlocks(successor, visitedBlocks, container);
+      }
+    }
+    return answer;
+  }
+
+  @CheckForNull
+  private Set<Block> localSuccessors(@Nonnull CFG.Block block, Map<Tree, CFGLoop> container) {
+    if (isStarting(block)) {
+      CFGLoop loop = container.get(block.terminator());
+      if (loop == null) {
+        loop = new CFGLoop(block, container);
+      }
+      Set<Block> loopSuccessors = new HashSet<>(loop.successors);
+      if (block.falseBlock() != null) {
+        loopSuccessors.add(block.falseBlock());
+      } else if (block.trueBlock() == null) {
+        // Special case where no condition is given in FOR loop: only one successor specified
+        return null;
+      }
+      return loopSuccessors;
+    }
+    return block.successors();
+  }
+
+  private boolean isBreak(@Nonnull CFG.Block block) {
+    return block.terminator() != null && block.terminator().is(Tree.Kind.BREAK_STATEMENT);
+  }
+
+  private void collectWaysOut(Map<Tree, CFGLoop> container) {
+    for (CFG.Block block : blocks) {
+      if (isStarting(block)) {
+        CFGLoop innerLoop = container.get(block.terminator());
+        successors.addAll(innerLoop.successors());
+      } else {
+        successors.addAll(block.successors());
+      }
+    }
+    successors.removeAll(blocks);
+    successors.remove(startingBlock);
+  }
+
+  @VisibleForTesting
+  static List<CFG.Block> sorted(Collection<CFG.Block> collection) {
+    List<CFG.Block> answer = new ArrayList<>(collection);
+    Collections.sort(answer, new Comparator<CFG.Block>() {
+      @Override
+      public int compare(CFG.Block o1, CFG.Block o2) {
+        // Use order of IDs
+        return Integer.compare(o1.id(), o2.id());
+      }
+    });
+    return answer;
+  }
+
+  private static boolean isStarting(CFG.Block block) {
+    Tree terminator = block.terminator();
+    return terminator != null && terminator.is(Tree.Kind.FOR_STATEMENT, Tree.Kind.WHILE_STATEMENT, Tree.Kind.DO_STATEMENT);
+  }
+
+  public static Map<Tree, CFGLoop> getCFGLoops(CFG cfg) {
+    Map<Tree, CFGLoop> cfgLoops = new HashMap<>();
+    for (CFG.Block block : cfg.blocks()) {
+      if (CFGLoop.isStarting(block)) {
+        Tree terminator = block.terminator();
+        if (!cfgLoops.containsKey(terminator)) {
+          // Newly created loops must register themselves to the map inside the constructor
+          new CFGLoop(block, cfgLoops);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either remove this useless object instantiation of class "CFGLoop" or use it [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1848)